sim/src/main/cc/bridges/TracerVModule.cc

```diff
@@ -0,0 +1,265 @@
+    const unsigned timeout = 10000 + s;
+    bool was_done = false;
+    for (unsigned i = 0; i < timeout; i++) {
+
+      for (auto &bridge : bridges) {
+        bridge->tick();
+      }
+      tracerv->tick();
+
+      if (simif->done()) {
+        was_done = true;
+        break;
+      }
+    }
+
+    if (!was_done) {
+      std::cout << "Hit timeout of " << timeout
+                << " tick loops afer a requested " << s << " steps"
+                << std::endl;
+    }
+
+    return was_done;
+  }
+
+  std::vector<uint64_t> expected_cycle;
+  std::vector<uint64_t> expected_iaddr;
+
+  std::vector<uint64_t> got_cycle;
+  std::vector<uint64_t> got_iaddr;
+
+  void got_instruction(const uint64_t cycle, const uint64_t pc) {
+    // std::cout << "cycle: " << cycle << " pc: " << pc << std::endl;
+    got_cycle.emplace_back(cycle);
+    got_iaddr.emplace_back(pc);
+  }
+
+  int simulation_run() override {
+    if (!tracerv) {
+      std::cout << "tracerv was never set" << std::endl;
+    }
+
+    // set the callback to capture traced instructions
+    tracerv->set_callback(std::bind(&TracerVModule::got_instruction,
+                                    this,
+                                    std::placeholders::_1,
+                                    std::placeholders::_2));
+
+    // Reset the DUT.
+    peek_poke->poke("reset", 1, /*blocking=*/true);
+    simif->take_steps(1, /*blocking=*/true);
+    peek_poke->poke("reset", 0, /*blocking=*/true);
+    simif->take_steps(1, /*blocking=*/true);
+
+    // the value of the first cycle as returned from TracerV
+    const uint64_t cycle_offset = 3;
+
+    // modified as we go
+    uint64_t e_cycle = cycle_offset;
+
+    const unsigned tracerv_width = tracerv->max_core_ipc;
+
+    // load MMIO and capture expected outputs
+    auto load = [&](std::vector<uint64_t> iad, std::vector<bool> bit) {
+      assert(iad.size() == bit.size());
+      const auto sz = iad.size();
+      for (unsigned i = 0; i < sz; i++) {
+        // std::cout << "loading " << i << " with " << iad[i] << "," << bit[i]
+        // << std::endl;
+        peek_poke->poke(namei(i), iad[i], true);
+        peek_poke->poke(namev(i), bit[i], true);
+
+        // calculate what TraverV should output, and save it
+        if (bit[i]) {
+          expected_cycle.emplace_back(e_cycle);
+          expected_iaddr.emplace_back(iad[i]);
+        }
+      }
+      e_cycle++;
+    };
+
+    // loop over tests. choose random valids with a simple pattern of iaddr
+    // load into MMIO, and tick the system
+    for (unsigned test_step = 0; test_step < get_total_trace_tests();
+         test_step++) {
+      const uint64_t pull = simif->rand_next(tracerv_width);
+
+      auto pull_iaddr = get_iaddrs(test_step, tracerv_width);
+      auto pull_bits = get_contiguous(pull, tracerv_width);
+
+      load(pull_iaddr, pull_bits);
+      steps(1);
+    }
+
+    auto [final_iaddr, final_valid] = get_final_values(tracerv_width);
```

Comment:
If the return type is a non-primitive, it's best to capture it as a const-ref: `const auto &[final_iaddr, final_valid]`